File: sandbox/gumball/gumball.kt

class GumballMachine : IGumballMachineCrank {

	val model  = "M1V192019"
		get() = "Model# ${field}"
	val serial = "123997919189"
		get() = "Serial# ${field}"

	var coinValue : Int = 0
	var cost = 0  // configure at setup	
	var count : Int 

    init {
        count = 100
    }

    fun configCost( theCost : Int ) {
    	cost = theCost
    }

	override fun insert( value : Int ) { 
		var newValue = coinValue + value
		println( "${newValue} > ${cost}?")
		if ( newValue > cost ) {
			println( "There's already enough money!" ) 
		}
		else {
			coinValue = newValue 
		}
	}	

	override fun eject() {
		if ( coinValue < cost ) {
			println( "Not Enough Money!" ) 
			return 
		}
		if ( count > 0 ) {
			count-- 
			println( "Gumball Ejected!" ) 
			println( "Change Return: " + (coinValue - cost) + " cents." ) 
			coinValue = 0 
		}
		else {
			println( "Sorry! Out of Gumballs." ) 
		}
	}
    
	fun debug() {
		println( "${model} ${serial} cnt:${count} value:${coinValue} cost:${cost}" )
	}

}

interface IGumballMachineCrank {

	fun insert( value : Int )
	fun eject()

	fun insertQuarter() { 
		insert( 25 ) 
	}

	fun turnTheCrank() {
		eject()
	}

}

fun GumballMachine.insertCoin( value : Int ) {
	coinValue += value 
	if (coinValue >= cost)
		eject()  
}

fun GumballMachine.returnCoins( ) {
	println( "Change Return: " + coinValue  + " cents." ) ;
	coinValue = 0
}

fun main() {

	var m = GumballMachine()
	m.configCost(50)
	m.debug()

	// Crank Test
	m.insertQuarter()
	m.debug()
	m.turnTheCrank()
	m.debug()
	m.insertQuarter()
	m.debug()
	m.turnTheCrank()
	m.debug()

	// Slot Test 
	m.insertCoin(25)
	m.debug()
	m.insertCoin(100)	
	m.debug()


}

File: sandbox/gumball/gumball.rb

module CrankModel
  def insertQuarter
    insert( 25 )
  end

  def turnTheCrank
    eject()
  end
end

module SlotModel
  def insertCoin( value )
    @coinValue = @coinValue + value 
    if @coinValue >= @cost
      eject()
    end
  end

  def returnCoins
    puts "Change Return: #{@coinValue}  cents." 
  end
end


class GumballMachine

  def initialize(c)
 	@model = "M1V192019"
 	@serial = "123997919189" 
 	@coinValue = 0 	
    @cost = c
    @count = 100
  end

  def configCost( theCost )
  	cost = theCost
  end

  def insert( value )
    newValue = @coinValue + value
  	puts "#{newValue} > ${@cost}?"
  	if newValue > @cost
  		puts "There's already enough money!"
  	else
  		@coinValue = newValue 
  	end
  end

  def eject()
  	if @coinValue < @cost 
  		puts "Not Enough Money!"
  	elsif @count > 0
  		@count = @count - 1
  		puts "Gumball Ejected!"
  		puts "Change Return:  #{@coinValue - @cost} cents."
  		@coinValue = 0 
  	else
  		puts "Sorry! Out of Gumballs."
  	end
  end

  def debug
    puts "#{@model} #{@serial} val:#{@coinValue} cost:#{@cost} cnt:#{@count}"
  end

  include CrankModel
  include SlotModel

end


m = GumballMachine.new(50)
m.configCost(50)
m.debug()

## Crank Test
m.insertQuarter()
m.insertQuarter()
m.turnTheCrank()


## Slot Test 
m.insertCoin(25)
m.insertCoin(100)	

